### PR TITLE
chore(storage): consolidate config and options validation functions

### PR DESCRIPTION
**Pull request by AllanZhengYP**
#### Description of changes

Add a `resolveS3ConfigAndInput` function as lowest common denominator for S3 config and input validation and resolution function.

#### Description of how you validated changes
Unit test

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts [hunk 1]
```diff
@@ -10,6 +14,11 @@ import { resolvePrefix as defaultPrefixResolver } from '../../../utils/resolvePr
 import { S3Options } from '../types';
 import { ResolvedS3Config } from '../types/options';
 
+type ApiOptions = Omit<S3Options, 'accessLevel' | 'targetIdentityId'> & {
+	accessLevel?: StorageAccessLevel;
+	targetIdentityId?: string;
+};
```
Comment (jimblanc) [inline packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:20]: Why are we omitting things and then adding them right back?

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts [hunk 2]
```diff
@@ -30,7 +39,8 @@ type ResolvedS3ConfigAndInput = {
  * @internal
  */
 export const resolveS3ConfigAndInput = async (
-	s3Options?: S3Options
+	amplify: AmplifyClassV6,
```
Comment (jimblanc) [inline packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:42]: Nit: would prefer to call this something else to make it more clear it's not the imported instance, but probably fine

**Review by jimblanc — COMMENTED** [review]
Looks good, thanks for doing this! Some questions & nits below

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts (hunk 1 above)
Comment (AllanZhengYP) [reply packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:20]: @jimblanc I was too lazy to create a new type. I will do that instead. The reason is that `S3Options` type is too tight -- it does not allow `targetIdentityId` if `accessLevel` is not `protected`. The input of the validation function should be more permissive than that

**Review by jimblanc — DISMISSED** [review]

**Review by jimblanc — APPROVED** [review]

File: packages/storage/__tests__/utils/resolveS3ConfigAndInput.test.ts [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+	return {
+		...core,
+		Amplify: {
+			...core.Amplify,
+			getConfig: jest.fn(),
+			libraryOptions: {},
+		},
+		fetchAuthSession: jest.fn(),
+	};
+});
```
Comment (kvramyasri7) [inline packages/storage/__tests__/utils/resolveS3ConfigAndInput.test.ts:24]: Nit:
```suggestion
jest.mock('@aws-amplify/core', () => ({
	fetchAuthSession: jest.fn(),
	Amplify: {
		getConfig: jest.fn(),
	},
}));
```

File: packages/storage/src/providers/s3/apis/uploadData/multipart/uploadHandlers.ts [hunk 4]
```diff
@@ -1,6 +1,8 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
 
+import { Amplify, StorageAccessLevel } from '@aws-amplify/core';
```
Comment (kvramyasri7) [inline packages/storage/src/providers/s3/apis/uploadData/multipart/uploadHandlers.ts:4]: Just a question: Why are we using `Amplify` and `AmplifyClassV6` importing from different places ? Is there any specific reason 🤔?

**Review by kvramyasri7 — DISMISSED** [review]

File: packages/storage/src/providers/s3/apis/uploadData/multipart/uploadHandlers.ts (hunk 4 above)
Comment (AllanZhengYP) [reply packages/storage/src/providers/s3/apis/uploadData/multipart/uploadHandlers.ts:4]: Amplify is the singleton instance, `AmplifyClassV6` class/type name of the Amplify singleton. We use `AmplifyClassV6` as a type. The Amplify instance can be client-side instance, or SSR instance

**Review by jimblanc — APPROVED** [review]

**Review by kvramyasri7 — APPROVED** [review]
lgtm

**Pull request merged**